### ISSUE TITLE
Memory leak when using apollo-client in an iframe.

### ISSUE THREAD

**Issue by KutnerUri** [apollographql/apollo-client#7654]
Hello,

It seems that closing an iframe, while Apollo is loading a request inside, together with user interaction, is causing the iframe to crush.
Sounds niche, but my app is heavily reliant on iframes to operate and separate user code from our main site.

Some hints I found trying to debug it:
* It happens only in Chrome. (which most our users have)
* Happens in standalone projects and different GQL servers.
* It does not happen if I implement the request with `fetch` and `useState()` instead of `@apollo/client`
* It only happens together with user interaction.
* It happens less in "production" mode. (Might be preventable by some Webpack optimization?)

**Intended outcome:**
Apollo client should gracefully shut itself down and free resources when closing. (maybe using the window onunload event)

**Actual outcome:**
Chrome freezes. :(
No errors are emitted to the log.

<img width="800" alt="Screen Shot 2021-02-04 at 20 21 02" src="https://user-images.githubusercontent.com/5400361/106937534-8bfd8d00-6726-11eb-9010-f72f5693f28f.png">

**How to reproduce the issue:**
I managed to reproduce the bug using the code in this gist:
https://gist.github.com/KutnerUri/5b65a2893c30a40646fe99726847a072
Paste it in a `create-react-app` to reproduce it yourself. (I used React 16)

![Kapture 2021-02-04 at 20 24 53](https://user-images.githubusercontent.com/5400361/106938023-1e9e2c00-6727-11eb-8c48-b03a2e36006b.gif)

Does not happen using fetch and react state:
![Kapture 2021-02-04 at 20 28 45](https://user-images.githubusercontent.com/5400361/106938411-a1bf8200-6727-11eb-8e47-091aa3fa6fa6.gif)

I could not make a live website using CodeSandbox, as it didn't allow iframes.

**Versions**
```
  @apollo/client:  3.3.6
  react: 16.13.1
  react-dom: 16.13.1

  System:
    OS: macOS 11.1
  Binaries:
    Node: 12.19.0 - ~/.nvm/versions/node/v12.19.0/bin/node
    Yarn: 1.22.10 - ~/.nvm/versions/node/v12.19.0/bin/yarn
    npm: 6.14.11 - ~/.nvm/versions/node/v12.19.0/bin/npm
  Browsers:
    Chrome: 88.0.4324.146
    Edge: 88.0.705.56
    Firefox: 84.0.2
    Safari: 14.0.2
```

**Comment by KutnerUri** [comment]
The bug seems to be solved in React 17, and is indeed cause by some loose `setState()` (or similar) after unmouting.
See here - https://github.com/facebook/react/pull/19220
And the underlying bug in chrome here - https://bugs.chromium.org/p/chromium/issues/detail?id=956832

I still think it's relevant and fixable since some projects still run on React 16 (React Native Web, for example)
For these projects I've created a super slim [hook](https://www.npmjs.com/package/@teambit/base-ui.hooks.use-graphql-light) (see docs [here](https://bit.dev/teambit/base-ui))

Otherwise, I'm pushing React 17 to my projects and hope it's the last I see of it. :)

**Comment by hwillson** [comment]
Let us know if this is still a concern with `@apollo/client@latest` - thanks!

**Comment by KutnerUri** [comment]
Most certainly still happening. @hwillson 
I just checked it now with `@apollo/client: 3.3.20`, and `react: 16.14.0`

**Comment by hwillson** [comment]
Thanks @KutnerUri - any chance you could share a runnable reproduction that shows this, to make it easier for maintainers to take a look? Also, would you mind trying with `@apollo/client@beta`?

**Comment by KutnerUri** [comment]
yes - copy this to a CRA's app.js, and click on `toggle` repeatedly.
gist.github.com/KutnerUri/5b65a2893c30a40646fe99726847a072

I know this is kind of a stress test, but it represents a real world scenario.
It seems that updating state while the window/iframe is closing exploits some bug in Chrome, and making it crush.

**Comment by KutnerUri** [comment]
I can confirm it still happens in `@apollo/client` version `3.4.0-rc.6`

**Comment by brainkim** [comment]
Here’s a reproduction https://github.com/apollographql/react-apollo-error-template/tree/memory-leak-iframe

**Comment by phryneas** [comment]
@jbhanu-thoughtspot With [brainkim](https://github.com/brainkim)'s reproduction up there I can verify that it doesn't happen with the Apollo Client and React versions you are naming, so you are likely encountering a different bug.

If you can reproduce the problem you are having with current versions of Apollo Client and React, please open a new issue with reproduction and I'd be happy to take a look, but I'm sorry: generally we cannot support versions that are that outdated - Apollo Client 3.4.13 was released back in 2021 and we have released 9 minor versions and countless patch versions since then and are currently working on a 4.0 release.  
The `useQuery` hook that was causing this bug has fully been rewritten at least two times since.

I do suspect that this is a React bug, since Apollo Client doesn't interact that deeply with browser APIs, but I am also not very hopeful that the React team will be able to help you here - React 17 was released back in 2020 and mostly as a temporary solution to ease migration to React 18.
The current version of React is 19, and they stopped shipping bug fixes for React 18 around June 2022 - the only releases after that were adding migration warnings and deprecation messages.

I did a short search and you could be encountering this React bug: https://github.com/facebook/react/pull/19220 that was [fixed in React 18.2](https://github.com/facebook/react/blob/main/CHANGELOG.md#1820-june-14-2022)
As this only seems to happen in development mode, maybe you could try and see if you accidentally ship dev builds instead of prod builds?

As I cannot reproduce this bug in newer versions of Apollo Client, I'm going to close this issue - but as I said, please don't hesitate to open a new issue if you can reproduce this or a similar bug in current versions of Apollo Client.
